### PR TITLE
Generate a warning when a commit reference is used

### LINKED ISSUE #4485: Add warning to `validate` if a version constraint contains a commit reference

### PR DESCRIPTION
**Pull request by jeroenseegers**
Closes #4485

### CODE REVIEW

File: tests/Composer/Test/Util/ConfigValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ */
+class ConfigValidatorTest extends TestCase
+{
+    /**
+     * Test ConfigValidator warns on commit reference
+     */
+    public function testConfigValidatorCommitRefWarning()
+    {
+        $configValidator = new ConfigValidator(new NullIO());
+        $reflection      = new \ReflectionClass(get_class($configValidator));
```
Comment (alcohol) [inline tests/Composer/Test/Util/ConfigValidatorTest.php]: I personally don't like aligning assignment operators cause it makes for messy diffs if you ever add or subtract a line (and end up having to realign everything).

File: src/Composer/Util/ConfigValidator.php [hunk 2]
```diff
@@ -111,25 +112,28 @@ public function validate($file, $arrayLoaderValidationFlags = ValidatingArrayLoa
...
-
             if (!empty($requireOverrides)) {
                 $plural = (count($requireOverrides) > 1) ? 'are' : 'is';
-                $warnings[] = implode(', ', array_keys($requireOverrides)). " {$plural} required both in require and require-dev, this can lead to unexpected behavior";
+                $this->warnings[] = implode(', ', array_keys($requireOverrides)). " {$plural} required both in require and require-dev, this can lead to unexpected behavior";
             }
         }
 
+        // check for commit references
+        $this->checkForCommitReferences($manifest['require']);
```
Comment (alcohol) [inline src/Composer/Util/ConfigValidator.php]: I think it'll be fine if you just merge both arrays and iterate here. No need to make warnings a private property then either.

**Comment by alcohol** [comment]
Thanks.

File: src/Composer/Util/ConfigValidator.php [hunk 3]
```diff
@@ -124,6 +124,19 @@ public function validate($file, $arrayLoaderValidationFlags = ValidatingArrayLoa
             }
         }
 
+        // check for commit references
+        $require = isset($manifest['require']) ? $manifest['require'] : array();
+        $requireDev = isset($manifest['require-dev']) ? $manifest['require-dev'] : array();
+        $packages = array_merge($require, $requireDev);
+        foreach ($packages as $package => $version) {
+            if (preg_match('/#/', $version) === 1) {
```
Comment (staabm) [inline src/Composer/Util/ConfigValidator.php]: using `strpos' will be way faster

Comment (alcohol) [inline src/Composer/Util/ConfigValidator.php]: Thank god we don't run `composer validate` a thousand times per second.

**Pull request merged**